src/lucky/flash.cr

```diff
@@ -16,78 +16,113 @@ module Lucky
       def call(context)
         call_next(context)
       ensure
-        flash = context.flash.not_nil!
-        context.session[PARAM_KEY] = flash.to_session
+        context.session[PARAM_KEY] = context.flash.to_json
       end
     end
 
-    class Store < Hash(String, String)
-      def self.from_session(json)
-        flash = new
-        values = JSON.parse(json)
-        values.each { |k, v| flash[k.to_s] = v.to_s }
-        flash
+    class Store
+      @next_key_is_for_now : Bool = false
+
+      forward_missing_to all_messages
+
+      def self.from_session(session : Lucky::Session::AbstractStore) : Store
+        json = JSON.parse session.fetch(Lucky::Flash::Handler::PARAM_KEY, "{}")
+        new.tap do |flash|
+          json.each do |key, value|
+            flash.now[key.to_s] = value.to_s
+          end
+        end
       rescue e : JSON::ParseException
         new
       end
 
       def initialize
-        @read = [] of String
-        @now = [] of String
-        super
+        @now = {} of String => String
+        @next = {} of String => String
+      end
+
+      private def all_messages
+        @next.merge(@now)
       end
 
       def fetch(key : String)
-        @read << key
-        super
+        @now.fetch(key)
       end
 
       def fetch(key : Symbol)
         fetch key.to_s
       end
 
+      def fetch(_not_accepted : T) forall T
+        {% raise "Flash key must be a Symbol or String, got #{T}" %}
+      end
+
+      def keep_all
+        @now.each do |key, value|
+          @next[key] ||= value
+        end
+      end
+
+      def []?(key : Symbol | String)
+        all_messages[key.to_s]?
+      end
+
+      def [](_not_accepted : T) forall T
+        {% raise "Flash key must be a Symbol or String, got #{T}" %}
+      end
+
+      def [](key : Symbol | String)
+        all_messages[key.to_s]
+      end
+
+      def []?(_not_accepted : T) forall T
```

Comment:
These make it so anything other than a Symbol or String raises a helpful message